BlockChainRunner.java

```diff
@@ -0,0 +1,354 @@
+                ));
+                continue;
+            }
+
+            for (TransactionInput pendingInput : pendingTransaction.getInputs()) {
+                String sourceTxId = pendingInput.getTxId();
+                int sourceOutputIndex = pendingInput.getIndex();
+
+                Transaction sourceTransaction = getTransactionFromBlockChainById(sourceTxId);
+                TransactionOutput sourceOutput = sourceTransaction.getOutputs().get(sourceOutputIndex);
+
+                PublicKey sourcePk = sourceOutput.getDestPublicKey();
+                List<UnspentOutput> userUnspentOutputs = new ArrayList<>(unspentOutputMap.get(sourcePk));
+                Optional<UnspentOutput> unspentOutput = userUnspentOutputs.stream()
+                    .filter(output -> output.getTxId().equals(sourceTxId) && output.getOutputIndex() == sourceOutputIndex)
+                    .findFirst();
+
+                if (unspentOutput.isEmpty()) {
+                    log(String.format(
+                        "INFO: Invalid transaction %s: no unspent output present", pendingTransaction.getTxId()
+                    ));
+                    isValid = false;
+                    break;
+                }
+
+                sign.initVerify(sourceOutput.getDestPublicKey());
+                sign.update(sourceTransaction.getTxId().getBytes());
+                if (!sign.verify(pendingInput.getSignedTxId())) {
+                    log(String.format(
+                        "INFO: Invalid transaction %s: Invalid digital signature", pendingTransaction.getTxId()
+                    ));
+                    isValid = false;
+                    break;
+                }
+
+                mappingsToRemove.add(Pair.of(sourcePk, unspentOutput.get()));
+                transactionInputValue += sourceOutput.getValue();
+            }
+            if (!isValid) continue;
+
+            int transactionOutputValue = pendingTransaction.getOutputs()
+                .stream()
+                .mapToInt(TransactionOutput::getValue)
+                .sum();
+            if (transactionInputValue < transactionOutputValue) {
+                log(String.format(
+                    "INFO: Invalid transaction %s: More output that input", pendingTransaction.getTxId()
+                ));
+                isValid = false;
+            }
+
+            if (isValid) {
+                mappingsToRemove.forEach(
+                    mapping -> unspentOutputMap.removeMapping(mapping.getKey(), mapping.getValue())
+                );
+                transactionPool.add(pendingTransaction);
+                totalTransactionsPooled++;
+            }
+        }
+
+        pendingTransactions = new ArrayList<>();
+    }
+
+    private int getUnspentOutputValue(UnspentOutput output) {
+        Transaction transaction = getTransactionFromBlockChainById(output.getTxId());
+        if (isNull(transaction) || transaction.getOutputs().size() <= output.getOutputIndex()) {
+            return 0;
```

Comment:
A method should have only one exit point and suitably it should be the last statement in the method. 

It would be advisable to reform the code to have only one return statement.

Reply:
Having only a single return statement would require a long ternary which would hurt readability